### PR TITLE
Refactor SparseApplyFtrl CPU kernel into class

### PR DESCRIPTION
**Pull request by benbarsdell**
This is in preparation for adding a GPU implementation.

(This is the second split of https://github.com/tensorflow/tensorflow/pull/43299).

cc @nluehr @sanjoy 

Edit: This PR makes no functional change except for the indexing error message (which is changed to match that of SparseApplyKerasMomentumOp which has the same code structure).

### CODE REVIEW

**Review by Lukasz257 — APPROVED** [review]

File: tensorflow/core/kernels/training_ops.cc [hunk 1]
```diff
@@ -350,6 +350,171 @@ struct ApplyFtrlMultiplyLinearByLr<CPUDevice, T> {
...
+    auto l1_reg_adjust = std::max(std::min(linear, l1), -l1);
+    return (l1_reg_adjust - linear) / quadratic;
+  }
+}
+
+}  // namespace
+
+template <typename T, typename Tindex, bool has_l2_shrinkage>
+struct SparseApplyFtrl<CPUDevice, T, Tindex, has_l2_shrinkage> {
+  Tindex operator()(const CPUDevice& d, typename TTypes<T>::Matrix var_flat,
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:385]: Would it be too much work for you to explicitly document what the `operator()` does (i.e. "It computes the following expression: ...").

The current code isn't documented so I'll approve the change regardless, but if you already have the context in your head already then maybe adding some extra documentation won't be too difficult?

**Review by sanjoy — CHANGES_REQUESTED** [review]
Minor comment inline.

I'm assuming this PR is NFC (no functional change); if yes then please change the description to say so, if not, please say that as well.

**Comment by benbarsdell** [comment]
Description edited to describe NFC.

File: tensorflow/core/kernels/training_ops.cc (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc:385]: I'm afraid I don't have that context in my head, as I didn't need to grok all of the math to make the changes.

File: tensorflow/core/kernels/training_ops.cc [hunk 2]
```diff
@@ -2672,146 +2808,23 @@ class SparseApplyFtrlOp : public OpKernel {
...
+            // Note: Passing l2 as a placeholder when not has_l2_shrinkage (it
+            // will not be used).
+            has_l2_shrinkage ? l2_shrinkage->scalar<T>() : l2.scalar<T>(),
+            lr_power.scalar<T>(), grad.flat_outer_dims<T>(), indices_vec,
+            inner_dim, multiply_linear_by_lr_);
+    OP_REQUIRES(
+        ctx, bad_i < 0,
+        errors::InvalidArgument(
+            "indices", SliceDebugString(indices.shape(), bad_i), " = ",
+            indices_vec(bad_i), " is not in [0, ", var.dim_size(0), ")"));
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:2827]: IMO making this non local is harder to understand.  Can you have `operatior()` return a `Status` (`OK` or `InvalidArgument`) and use `OP_REQUIRES_OK` here?

**Review by sanjoy — CHANGES_REQUESTED** [review]
Minor comment inline, lgtm other than that.

File: tensorflow/core/kernels/training_ops.cc (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc]: Done

**Review by sanjoy — APPROVED** [review]
lgtm

**Pull request merged**